docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+The ``@overload``-decorated definitions must be followed by an overload
+implementation, which does not include an ``@overload`` decorator. Type
+checkers should report an error or warning if an implementation is missing.
+Overload definitions within stub files, protocols, and abstract base classes
+are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Type checkers should report
+an error if these conditions are not met.
+
+If one or more overloads are decorated with ``@final`` or ``@override`` but the
+implementation is not, an error should be reported.
+
+Overloads are allowed to use a mixture of ``async def`` and ``def`` statements
+within the same overload definition. Type checkers should desugar all
+``async def`` statements before testing for implementation consistency
+and overlapping overloads (described below).
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that it is consistent with all of its associated overload signatures.
+The implementation should accept all potential sets of arguments
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be assignable to the input signatures
+of all overloads, and the return type of all overloads should be assignable to
+the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+When a type checker checks the implementation for consistency with overloads,
+it should first apply any transforms that change the effective type of the
+implementation including the presence of a ``yield`` statement in the
+implementation body, the use of ``async def``, and the presence of additional
+decorators.
+
+
+Overlapping overloads
+^^^^^^^^^^^^^^^^^^^^^
+
+If two overloads can accept the same set of arguments, they are said
+to "partially overlap". If two overloads partially overlap, the return type
+of the latter overload should be assignable to the return type of the
+former overload. If this condition doesn't hold, it is indicative of a
+programming error and should be reported by type checkers::
```

Comment:
@erictraut In another comment thread you mentioned that even in the presence of an "overlapping overload" error, resolution of the overload should still be possible and follow these rules as written. This makes sense, but it wasn't immediately obvious to me on my first read-through; there are of course cases where if there is a type error we have to just "give up" and fall back to `Any`. Maybe this is something we want to state explicitly? (That an overlapping-overloads error does not imply giving up on the overload resolution.)